### PR TITLE
2480 - Setup public database migrations

### LINKED ISSUE #2480: Setup public database migrations
### Public schema
`src/tools/migrations/public`

- [x] Setup migration for public schema and tables (DDL) on empty database
  - [x] Create a file with public.ddl (check src/server/repository/assessment/assessment/getCreateSchemaDDL.ts)
- [x] Insert default admin user (admin@admin.com:admin1234)
- [x] Create the command: `yarn migrations-public:create`
- [x] Create the command: `yarn migrations-public:run`
- [ ] Run public migrations on every start of the application
- [x] Create a migration step to initialise existing application
- [x] Remove old migration (not migration-step)

### PR DESCRIPTION
**Pull request by sorja**
## developers note

To test this branch, create a new database `fra-db-test` inside the docker:

```bash
# terminal
docker exec -it fra-db bash

# inside docker container
time createdb -U frap frap-dev-test

# to rerun migrations / drop db:
time dropdb -U frap frap-dev-test #drop db
time createdb -U frap frap-dev-test #create db
```

Update the `.env` 

```config
PGDATABASE=frap-dev-test
```

- [x] Update `public.migration` created on auto populate
- [x] Admin user creation should be under tools 

Reviewing the PR:
It is suggested to hide the deleted files in the PR view
Screenshot 2024-09-16 at 12 34 54

### CODE REVIEW

**Comment by minotogna** [comment]
@sorja can we name this pr as its issue ?

**Comment by sorja** [comment]
@minotogna 

Is this fine to avoid running public-migrations in current instances?

```
src/test/migrations/steps/20240909101347-step-exclude-public-migrations.ts
```

Or should we merge this file to `dev` before this branch?

(Note: I will update it when I have all migrations in place that are needed for this main branch)

**Comment by minotogna** [comment]
> @minotogna 
> 
> 
> 
> Is this fine to avoid running public-migrations in current instances?
> 
> 
> 
> ```
> 
> src/test/migrations/steps/20240909101347-step-exclude-public-migrations.ts
> 
> ```
> 
> 
> 
> Or should we merge this file to `dev` before this branch?
> 
> 
> 
> (Note: I will update it when I have all migrations in place that are needed for this main branch)

Not sure I understand.
On current instances initial step should not run and we should have a migration step .

**Comment by sorja** [comment]
> > @minotogna
> > Is this fine to avoid running public-migrations in current instances?
> > ```
> > 
> > src/test/migrations/steps/20240909101347-step-exclude-public-migrations.ts
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Or should we merge this file to `dev` before this branch?
> > (Note: I will update it when I have all migrations in place that are needed for this main branch)
> 
> Not sure I understand. On current instances initial step should not run and we should have a migration step .

We discussed yesterday:
- Running migration on startup would run the public migrations before the steps

**Comment by minotogna** [comment]
> > > @minotogna
> 
> > > Is this fine to avoid running public-migrations in current instances?
> 
> > > ```
> 
> > > 
> 
> > > src/test/migrations/steps/20240909101347-step-exclude-public-migrations.ts
> 
> > > ```
> 
> > > 
> 
> > > 
> 
> > >     
> 
> > >       
> 
> > >     
> 
> > > 
> 
> > >       
> 
> > >     
> 
> > > 
> 
> > >     
> 
> > >   
> 
> > > Or should we merge this file to `dev` before this branch?
> 
> > > (Note: I will update it when I have all migrations in place that are needed for this main branch)
> 
> > 
> 
> > Not sure I understand. On current instances initial step should not run and we should have a migration step .
> 
> 
> 
> We discussed yesterday:
> 
> - Running migration on startup would run the public migrations before the steps

I don't remember we said that , it we didn't understand each other then .
The migration steps are executed manually . We decide the order .
I think in current instances , the steps gets executed before the public migrations , otherwise this logic doesn't work.

File: src/server/repository/public/ddl/getCreatePublicSchemaDDL.ts [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        on update no action on delete no action,
+      foreign key (cycle_uuid) references ${schemaName}.assessment_cycle (uuid)
+        on update cascade on delete cascade,
+      foreign key (user_id) references ${schemaName}.users (id)
+        on update cascade on delete cascade
+    );
+    create index if not exists idx_activity_log_cycle_message on ${schemaName}.activity_log using btree (cycle_uuid, message);
+    create index if not exists idx_activity_log_filtering on ${schemaName}.activity_log using btree (assessment_uuid, cycle_uuid, country_iso, message, time);
+
+    create table if not exists ${schemaName}.migration_steps (
```
Comment (sorja) [inline src/server/repository/public/ddl/getCreatePublicSchemaDDL.ts]: Not sure if this should be here, or should the script check and create if not exists?

File: src/server/repository/public/ddl/getCreatePublicSchemaDDL.ts [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+      id serial primary key,
+      name character varying(255) not null unique,
+      run_on timestamp without time zone not null default now()
+    );
+
+    -- This table is created separately, but left here for consistency
+    create table if not exists ${schemaName}.migrations (
```
Comment (sorja) [inline src/server/repository/public/ddl/getCreatePublicSchemaDDL.ts]: Same as above

Not sure if this should be here, or should the script check and create if not exists?


They seem like helper tables that are not directly related to our application
(I would even introduce some separate schema or something to separate them more)

File: src/tools/migrations/public/index.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import { DB } from 'server/db'
+import { Logger } from 'server/utils/logger'
+
+const client = DB
+let migrationSteps: Array<string>
+let previousMigrations: Array<string> = []
+const executedMigrations: Array<string> = []
+
+// TODO: Move to migration step / schema initialisation
+const tableDDL = `
```
Comment (sorja) [inline src/tools/migrations/public/index.ts]: This is also related to previous comments

File: src/server/repository/public/ddl/getCreatePublicSchemaDDL.ts (hunk 2 above)
Comment (minotogna) [reply src/server/repository/public/ddl/getCreatePublicSchemaDDL.ts]: interesting. should we have a separate schema with all migrations ?

Comment (sorja) [reply src/server/repository/public/ddl/getCreatePublicSchemaDDL.ts]: For me, it would make sense, let me know

Comment (minotogna) [reply src/server/repository/public/ddl/getCreatePublicSchemaDDL.ts]: Ok @sorja , it is cleaner with the migrations schema .

With two tables atm : public and steps 

In the migrations (public) let's add a step that initializes the schema and tables .

In the step that runs on current instances , move steps and initialize the public table .

What do you think ?

Comment (sorja) [reply src/server/repository/public/ddl/getCreatePublicSchemaDDL.ts]: My question is, what would be logical place to create the `migrations.public` and `migration.step` tables?

When running the first migration of the migration script, we don't have this table on an empty database and it relies on it.

So I would suggest always running `create table if not exists` before running the script, as it is ran quite rarely in the end, so it doesn't matter to check

Let me know

Comment (sorja) [reply src/server/repository/public/ddl/getCreatePublicSchemaDDL.ts]: I am referring to this way of doing it:
https://github.com/openforis/fra-platform/pull/3904/files#diff-c0946e94fffdfd9287c7d54e5dd3cb24cfff140c35caef52ae54652767ceeff3R17

Comment (minotogna) [reply src/server/repository/public/ddl/getCreatePublicSchemaDDL.ts]: it  must run before any migrations and on any new instances only.ok?

Comment (sorja) [reply src/server/repository/public/ddl/getCreatePublicSchemaDDL.ts]: Yes, it is currently (this pr) implemented as so, I will update the schema and schema creation in the creation of table
👍

**Review by minotogna — APPROVED** [review]
